File: 020-valid-parentheses.cpp
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for (char c : s) {
            switch (c) {
                case '(': stack.push(')'); break;
                case '[': stack.push(']'); break;
                case '{': stack.push('}'); break;
                default:
                    if (stack.empty() || stack.top() != c) return false;
                    stack.pop();
            }
        }
        return stack.empty();
    }
};

#include <iostream>

using namespace std;

int main()
{
    cout << Solution().isValid("([{}])") << endl;
    return 0;
}

File: 020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack, map = [], { ')':'(', ']':'[', '}':'{' }
        for c in s:
            if c in map.values():
                stack.append(c)
            else:
                if stack == [] or map[c] != stack.pop():
                    return False
        return stack == []
